Parrec sort fix

grlee77 opened the pull request:
This PR is an attempt to address #408.

I added a new boolean option upon loading a .PAR/.REC file called `strict_sort`.  If it is set to `True` the `get_sorted_slice_indices` method of `PARRECHeader` uses several additional keys in the call to `np.lexsort`.  `strict_sort=False` corresponds to the previously existing sort mode.  This option is made available via a `--strict_sort` command line option to `parrec2nii` as well.

In general, I think it is probably preferable to use `strict_sort=True` in all cases, but I have left the default as `False` for now to ensure backwards compatibility.  Perhaps after a deprecation/testing period we could make `True` the new default.  In particular, it may not be true that for any given DTI scan the b-vals/vectors would be sorted in exactly the same order as previously (If the user is using the b-vals/vectors as exported by `parrec2nii` this should not be an issue, but could cause problems if mixed with a previously stored table from the other approach).

There could be some debate as to the preferred order of priority for the sort keys.  I chose one that seemed logical to me.  As before, slices is the lowest priority so that dimension 3 of the data is still always slices and `is_full` is the highest so that partially acquired scans get moved to the end for possible discarding.  The sorting is a bit complicated, but I tried to add comments to make the reasoning clearer.  The full set of potential sort keys implemented is:
```
slice_nos
echos
cardiac phases
b-vectors
b-values
asl_labels
dynamics
vol_numbers()
is_full
```

grlee77 commented:
If I get a chance, I may scan a phantom to generate a couple of additional complex .PAR files to add to the tests.

matthew-brett commented:
That would be very good.

If you happen to get a chance to scan a phantom with (different) rotations about all three axes, that would always help a lot - we don't know what order PAR / REC uses for these at the moment.

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -631,10 +632,14 @@ def __init__(self, info, image_defs, permit_truncated=False):
             recording is detected.
+        strict_sort : bool, optional, keyword-only
+            If True, a larger number of header fields are used while sorting
+            the REC data array.
```
Perhaps mention that this can give a different order of slices.

matthew-brett commented on nibabel/tests/test_parrec.py:
```diff
@@ -517,9 +554,10 @@ class FakeHeader(object):
+    def __init__(self, shape, dtype, strict_sort=False):
         self._shape = shape
         self._dtype = np.dtype(dtype)
+        self.strict_sort = strict_sort
```
Is this actually used?

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -1008,13 +1018,71 @@ def get_sorted_slice_indices(self):
+                - call vol_is_full to identify potentially missing volumes
+                - call vol_numbers to assign unique volume numbers if for some
+                  reason the keys defined above don't provide a unique sort
+                  order (e.g. this occurs for the Trace volume in DTI)
```
Do we have an example in the tests where the keys don't produce a unique sort order?

grlee77 replied:
After looking into it a bit more closely, I don't think a non-unique order should ever occur (unless we are missing a sort key) and vol_numbers can be removed in the `strict_sort=True` case.  The Trace volume in DTI does actually have a different bval/bvec combination than any of the other volumes, I just didn't read the columns properly before.

One case where I found non-unique sort order was that I was missing a key for `image_type_mr` so when I saved a .PAR with multiple image types (real, imaginary, magnitude, phase), these didn't get sorted properly and there were multiple copies of the same slice per volume, etc unless vol_numbers was used.  However, if the `image_type_mr` key is added then there is a unique sort order again and it is unneeded.

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -1008,13 +1018,71 @@ def get_sorted_slice_indices(self):
+            echos = self.image_defs['echo number']
+
+            # try adding keys only present in a subset of .PAR files
+            try:
```
How about:

```
idefs = self.image_defs
asl_keys = (idefs['label_type'],) if 'label_type' in idefs else ()
```

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -1008,13 +1018,71 @@ def get_sorted_slice_indices(self):
+                    # only present for .PAR v4.1+
+                    bvals = self.image_defs['diffusion b value number']
+                    bvecs = self.image_defs['gradient orientation number']
+                except:
```
Very good to avoid bare `except` statements - can catch syntax errors, control-C, etc.   How about something like:

```
def get_def(name):
    return self.image_defs[name] if name in self.image_defs else None

bvals = get_def('diffusion b value number')
if bvals is None:
    bvals = get_def('diffusion_b_factor')
bvecs = get_def('gradient orientation number')
if bvecs is None:
    bvecs = tuple(get_def('diffusion'))
```

grlee77 replied:
Because `image_defs` is a labelled array and not a dictionary I had to modify `get_def` a bit:
```python
    def get_def(self, name):
        """ Return a single image definition field. """
        idef = self.image_defs
        return idef[name] if name in idef.dtype.names else None
```
otherwise the following problem occurs:
```FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
  return self.image_defs[name] if name in self.image_defs else None```

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -1008,13 +1018,71 @@ def get_sorted_slice_indices(self):
+                except:
+                    bvals = self.image_defs['diffusion_b_factor']
+                    # use hash to get a single sortable value per direction
+                    bvecs = [hash(tuple(
```
I think the only guarantee you have for a hash is that two values that compare equal will have the same hash : https://docs.python.org/2/library/functions.html#hash

There's no guarantee that two values that compare unequal will have a different hash though.  I think the sort order is arbitrary too.   Can't you just use the tuple itself for the sort?

grlee77 replied:
Yeah, this case is a bit tricky.  I haven't tested with tuples as the keys to `lexsort`, but maybe that would work.  A simpler option would be to just not support strict_sort with V4.0 PAR files.  As is, the sorting would behave differently for V4.0 where b-values are sorted by magnitude, but in V4.1 they are sorted using a "diffusion b value number" field which, I am not certain, but may instead be related to the order the b-values are scanned.

grlee77 replied:
I didn't have luck with tuple of tuples in lexsort, so I think I will handle this via:

```python
bvecs = _direction_numbers(self.get_def('diffusion'))
```
where I define the utility function
```python
def _direction_numbers(bvecs):
    """ Enumerate directions from an (N, 3) array of direction vectors.
    """
    cnt = 0
    bvec_dict = {}
    bvec_nos = np.zeros(len(bvecs))
    for i, bvec in enumerate(bvecs):
        bv = tuple(bvec)
        if bv not in bvec_dict:
            bvec_dict[bv] = cnt
            cnt += 1
        bvec_nos[i] = bvec_dict.get(bv)
    return bvec_nos
```

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -1008,13 +1018,71 @@ def get_sorted_slice_indices(self):
+            keys = (slice_nos, vol_numbers(slice_nos), np.logical_not(is_full))
+            sort_order = np.lexsort(keys)
+        else:
+            # Sort based on a larger number of keys.  This is more complicated
```
This function is getting long - maybe have a separate helper function for this?

matthew-brett commented:
Can you also add a test for `strict_sort` in the script, in `nibabe/test_scripts.py`?

I'm afraid I still haven't got my head round the sorting, but I'll play with the code soon to see if I can understand it.

grlee77 commented:
@matthew-brett 
As requested above (but unrelated to this PR), .PAR/.REC files corresponding to various oblique image orientations are made freely available here:
https://github.com/grlee77/parrec_oblique

I have not yet had the time to fully compare the affines produced by the scanner to those generated by `parrec2nii`.

grlee77 commented:
I tried to look at the details for the coveralls failure above, but I get the following error:
```
Source Not Available

The file "/home/travis/build/nipy/nibabel/venv/lib/python2.7/site-packages/nibabel/parrec.py" isn't available on github. Either it's been removed, or the repo root directory needs to be updated.
```

any ideas why that is?

effigies commented:
@grlee77 See #331.

matthew-brett commented:
Great - thanks for the data - I make a PR testing against it at #420

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -990,6 +1019,54 @@ def get_rec_shape(self):
+        is_full = vol_is_full(slice_nos[initial_sort_order],
+                              self.general_info['max_slices'])
+
+        # have to "unsort" is_full to match the other sort keys
```
This logic took me a bit to work out - is there a good test for it?

grlee77 replied:
Do you mean the `is_full` case specifically?  If so, I don't think I tested that aside from verifying that it still passes the existing tests.  None of the new data sets I added have missing data, although I suppose we could truncate one of the files if desired.

I don't have an exhaustive set of tests for all available sort keys, but I did collect a few .PAR files with 4-6 dimensions to use as tests.  These are the new .PAR files I added in tests/data:

- *T1_dual_echo.PAR*:  A 4D data set:  (x, y, z, echo)
- *T1_3cho_mag_real_imag_phase.PAR*:  A 5D data set:  (x, y, z, echo, image_type)
- *ASL_3D_Multiecho.PAR*:  A 6D data set:  (x, y, z, echo, 'label type', dynamics)
The tests make sure the orders are sorted as expected for these cases.  I did not collect any data with cardiac phases, for example.

One other test I can think of would be to manually scramble the order of the entries in one of the .PAR files and make sure the sort keys that come out at the end still match the unscrambled case.

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -329,6 +329,21 @@ def _process_image_lines(image_lines, version):
     return image_defs
 
 
+def _direction_numbers(bvecs):
```
Maybe put in a test for this helper function?  You could copy https://github.com/yarikoptic/nitest-balls1/blob/master/PARREC/DTI.PAR into the `tests/data` directory to get some sample bvecs.

grlee77 replied:
okay.  perhaps that was the source of the coveralls complaint

grlee77 replied:
Actually, no bvecs will be present in older V4 .PAR files so this `_direction_numbers` function is unnecessary and would never be called.  I will remove it.

I submitted a small PR to fix an existing bug when the 'diffusion' field isn't present in `image_defs`#426

grlee77 commented:
@matthew-brett 
I think this is now ready for review again.  The primary changes since last time you reviewed are:

- I removed the previously untested `_direction_numbers` routine since it was based on a 'diffusion' field in `image_defs` that isn't even present in the V4 .PAR files that would have been the only case to call this routine (see #426)
- In the tests for strict sort I made sure they pass even when the image_defs array is randomly sorted via `np.shuffle`.
- Since V4 .PAR files have no `gradient orientation number` key, I added `vol_numbers` back in as a key in the strict sort routine so the diffusion directions would get assigned unique volume numbers even for V4 .PAR files.  As was already the case for `vol_is_full` this has to be done after the initial stage of sorting.  In this way the sorting is robust even to random shuffling of the image info lines from the .PAR file.

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -725,19 +734,26 @@ def get_bvals_bvecs(self):
         return bvals, bvecs
 
+    def get_def(self, name):
+        """ Return a single image definition field. """
```
Add "or None if field not present" ?

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -725,19 +734,26 @@ def get_bvals_bvecs(self):
             (n_slices, n_vols), order='F')
-        # All bvals within volume should be the same
-        assert not np.any(np.diff(bvals, axis=0))
+        if not self.strict_sort:
```
Can you think of a better place for these checks?  Maybe there's a way to check these after doing the sorting?

grlee77 replied:
The check should also be performed in the strict sort case since these lines occur after the sort, so this change was a mistake!  I think these checks should then stay in this function.

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -990,6 +1006,62 @@ def get_rec_shape(self):
+              and add the result to the list of sort keys
+        """
+        # Sort based on a larger number of keys.  This is more complicated
+        # but works for .PAR files that get missorted by the above method
```
I guess it's not "the above method" now.

Maybe put the other sorting method into a `lax_sort` method, and refer to that?

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -990,6 +1006,62 @@ def get_rec_shape(self):
+        image_type = self.image_defs['image_type_mr']
+
+        # try adding keys only present in a subset of .PAR files
+        idefs = self.image_defs
```
Put this at top and use it for first set of `image_defs` references?

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -990,6 +1006,62 @@ def get_rec_shape(self):
+        asl_keys = (idefs['label type'], ) if 'label type' in \
+            idefs.dtype.names else ()
+
+        if not self.general_info['diffusion'] == 0:
```
Or:

```
if self.general_info['diffusion'] != 0:
```

matthew-brett commented on nibabel/parrec.py:
```diff
@@ -1008,13 +1085,18 @@ def get_sorted_slice_indices(self):
-        keys = (slice_nos, vol_numbers(slice_nos), np.logical_not(is_full))
-        # Figure out how many we need to remove from the end, and trim them
-        # Based on our sorting, they should always be last
+        if not self.strict_sort:
```
Maybe move into `_lax_sort_keys method` and do `keys = self._strict_sort_keys() if self.strict_sort else self._lax_sort_keys()`

matthew-brett commented:
Great - thanks for the changes - nearly there.

It's good to avoid backslashes for line continuation if possible - https://www.python.org/dev/peps/pep-0008/#id17

grlee77 commented:
I made the suggested changes above.

I also simplified the final lines of `_strict_sort_order`.  I was able to remove the "unsorting" entirely.  Now it is an initial sort as before, followed by a second sort that is just:

```initial_sort_order[np.lexsort((vol_nos, is_full))]```

The docstring was expanded to (hopefully!) better describe what is going on in the function`.

nibotmi commented:
:umbrella: The latest upstream changes (presumably #426) made this pull request unmergeable. Please resolve the merge conflicts.

grlee77 commented:
rebased to fix merge conflicts with #426

matthew-brett commented:
Thanks for your patience - it looks good - merging.

grlee77 commented:
Great!  This will be very useful for some of the data we acquire.
